lib/luaT/luaT.c

```diff
@@ -497,27 +497,77 @@ int luaT_classmodulename(const char *tname, char *module_name)
   return tname[n] == '.';
 }
 
-/* Lua only functions */
-int luaT_lua_newmetatable(lua_State *L)
+/* Method for determining name of class module */
+const char* luaT_classname(const char *tname)
 {
-  const char* tname = luaL_checkstring(L, 1);
-  char module_name[256];
-  int is_in_module = 0;
-  is_in_module = luaT_classmodulename(tname, module_name);
+  int idx;
+  int sz = strlen(tname);
+
+  for(idx = sz-1; idx >= 0 ; idx--)
+  {
+    if(tname[idx] == '.')
+      return tname+idx+1;
+  }
+  return tname;
+}
+
+void traverse_tables(lua_State *L, const char *tname, int *L_idx)
+{
+  /* Local variables */
+  char term[256];
+  char chars[] = {'.', '\0'};
+  const char *tname_full = tname; // used for error case
 
+  /* Get outermost table from Lua */
+  int n = strcspn(tname, chars);
+  strncpy(term, tname, n);
+  term[n] = '\0';
+  lua_getglobal(L, term);
+  tname  += n + 1;
+
+  /* Traverse hierarchy down to last table*/
+  n = strcspn(tname, chars);
+  while(n < strlen(tname))
+  {
+    if(!lua_istable(L, *L_idx)){
+      strncpy(term, tname_full, tname - tname_full - 1);
+      term[tname - tname_full] = '\0';
+      luaL_error(L, "while creating metatable %s: bad argument #1 (%s is an invalid module name)", tname_full, term);
+    }
+    strncpy(term, tname, n);
+    term[n] = '\0';
+    lua_getfield(L, *L_idx, term);
+    *L_idx += 1;
```

Comment:
If you remove the parent table after the `lua_getfield` call, you can avoid all the index computation.